pkg/efscreate/efs.go

```diff
@@ -0,0 +1,354 @@
+	service := ec2.New(sess)
+	efsClient := awsefs.New(sess)
+	return &EFS{
+		client:    service,
+		efsClient: efsClient,
+		infra:     infra,
+		subnetIDs: []string{},
+		resources: &ResourceInfo{},
+	}
+}
+
+func (efs *EFS) CreateEFSVolume(nodes *corev1.NodeList) (string, error) {
+	instances := efs.getInstanceIDs(nodes)
+	err := efs.getSecurityInfo(instances)
+	if err != nil {
+		return "", err
+	}
+	sgid, err := efs.createSecurityGroup()
+	if err != nil {
+		return "", err
+	}
+	efs.resources.securityGrouID = sgid
+	ok, err := efs.addFireWallRule(sgid)
+	if err != nil || !ok {
+		return "", fmt.Errorf("error adding firewall rule: %v", err)
+	}
+
+	fileSystemID, err := efs.createEFSFileSystem()
+	if err != nil {
+		return "", err
+	}
+	efs.resources.efsID = fileSystemID
+	mts, err := efs.createMountTargets(fileSystemID, sgid)
+	if err != nil {
+		return "", err
+	}
+	efs.resources.mountTargets = mts
+	log("successfully created file system %s", fileSystemID)
+	return fileSystemID, nil
+}
+
+func (efs *EFS) getInstanceIDs(nodes *corev1.NodeList) []string {
+	nodeIDs := sets.NewString()
+	for _, node := range nodes.Items {
+		//get providerID of the form aws:///us-west-2a/i-0304804a704fefb7d
+		instanceString := node.Spec.ProviderID
+		instanceStringArray := strings.Split(instanceString, "/")
+		if len(instanceStringArray) > 0 {
+			nodeIDs.Insert(instanceStringArray[len(instanceStringArray)-1])
+		}
+	}
+	return nodeIDs.List()
+}
+
+func (efs *EFS) createSecurityGroup() (string, error) {
+	infraID := efs.infra.Status.InfrastructureName
+	groupName := fmt.Sprintf(securityGroupNameFormat, infraID)
+	securityGroupInput := ec2.CreateSecurityGroupInput{
+		Description:       aws.String("for testing efs driver"),
+		GroupName:         aws.String(groupName),
+		VpcId:             &efs.vpcID,
+		TagSpecifications: efs.getTags(ec2.ResourceTypeSecurityGroup, groupName),
+	}
+	response, err := efs.client.CreateSecurityGroup(&securityGroupInput)
+	if err != nil {
+		return "", fmt.Errorf("error creating security group")
+	}
+	return *response.GroupId, nil
+}
+
+func (efs *EFS) getTags(resourceType string, resourceName string) []*ec2.TagSpecification {
+	var tagList []*ec2.Tag
+	tags := map[string]string{
+		"Name":                 resourceName,
+		efs.getClusterTagKey(): "owned",
+	}
+	for k, v := range tags {
+		tagList = append(tagList, &ec2.Tag{
+			Key: aws.String(k), Value: aws.String(v),
+		})
+	}
+	return []*ec2.TagSpecification{
+		{
+			Tags:         tagList,
+			ResourceType: aws.String(resourceType),
+		},
+	}
+}
+
+func (efs *EFS) getClusterTagKey() string {
+	return fmt.Sprintf(tagFormat, efs.infra.Status.InfrastructureName)
+}
+
+func (efs *EFS) addFireWallRule(groupId string) (bool, error) {
+	ruleInput := ec2.AuthorizeSecurityGroupIngressInput{
+		CidrIp:     aws.String(efs.cidrBlock),
+		GroupId:    aws.String(groupId),
+		IpProtocol: aws.String("tcp"),
+		ToPort:     aws.Int64(2049),
+		FromPort:   aws.Int64(2049),
```

Comment:
I'm curious how does it work. I am pretty sure that stunnels on nodes don't use source port `2049`. Is it ignored by AWS?

Reply:
I was confused about fromport argument as well. Apparently fromport..toport in this case is used to specify range and is not the source port of a typical firewall rule.